src/btree/bt_compact.c

```diff
@@ -242,11 +242,27 @@ __compact_walk_internal(WT_SESSION_IMPL *session, WT_REF *parent)
     /*
      * We could corrupt a checkpoint if we moved a block that's part of the checkpoint, that is, if
      * we race with checkpoint's review of the tree. Get the tree's flush lock which blocks threads
-     * writing pages for checkpoints, and hold it long enough to review a single internal page. Quit
-     * working the file if checkpoint is holding the lock, checkpoint holds the lock for relatively
-     * long periods.
+     * writing pages for checkpoints, and hold it long enough to review a single internal page. If
+     * checkpoint has the lock, check for a compact interrupt and cache stuck before trying again.
      */
-    WT_RET(__wt_spin_trylock(session, &S2BT(session)->flush_lock));
+    while ((ret = __wt_spin_trylock(session, &S2BT(session)->flush_lock)) == EBUSY) {
+        WT_RET(__wt_session_compact_check_interrupted(session));
+
+        if (__wt_cache_stuck(session))
+            WT_RET(EBUSY);
+
+        WT_STAT_CONN_INCR(session, session_table_compact_conflicting_checkpoint);
```

Comment:
I would do this first in the while loop as soon as we get EBUSY with the verbose message.

Reply:
Agreed